File: graphs/easy_depth_first_search.js
// Depth First Search

class Node {
  constructor(name) {
    this.name = name;
    this.children = [];
  }

  addChild(name) {
    this.children.push(new Node(name));
    return this;
  }

  // #1
  depthFirstSearch(array) {
    const stack = [ this ]

    while (stack.length) {
      const { name, children} = stack.pop()
      array.push(name)

      for (let i = children.length - 1; i >= 0; i--) {
        stack.push(children[i])
      }
    }

    return array
  }

  // #2 recursion
  depthFirstSearch(array) {
    const dfs = (root) => {
      if (root) {
        const { name, children } = root
        array.push(name)

        for (let child of children) {
          dfs(child)
        }
      }
    }

    dfs(this)

    return array
  }

  // #3
  depthFirstSearch(array) {
    // Write your code here.
    array.push(this.name)

    for (let child of this.children) {
      child.depthFirstSearch(array)
    }

    return array
  }
}

File: graphs/easy_depth_first_search.py
# Depth First Search

# 1
class Node:
  def __init__(self, name):
    self.children = []
    self.name = name

  def addChild(self, name):
    self.children.append(Node(name))
    return self

  def depthFirstSearch(self, array):
    stack = [ self ]

    while stack:
      curr = stack.pop()

      array.append(curr.name)

      # for i in range(len(curr.children)-1, -1,-1):
      #     stack.append(curr.children[i])
      for i in curr.children[::-1]:
        stack.append(i)

    return array


# 2
  def depthFirstSearch(self, array):
    # Write your code here.
    array.append(self.name)

    for child in self.children:
      child.depthFirstSearch(array)

    return array
